packages/build/src/plugins_core/deploy/buildbot_client.ts

```diff
@@ -0,0 +1,167 @@
+import net, { NetConnectOpts } from 'net'
+import { normalize, resolve, relative } from 'path'
+import { promisify } from 'util'
+
+import { pEvent } from 'p-event'
+
+import { addErrorInfo } from '../../error/info.js'
+import { runsAfterDeploy } from '../../plugins/events.js'
+import { addAsyncErrorMessage } from '../../utils/errors.js'
+
+/**
+ * Error types the Buildbot server sends, see:
+ * - https://github.com/netlify/buildbot/blob/934981c85c779610e15ce549846b0bea212b283e/bot/deploy.go#L107
+ * - https://github.com/netlify/buildbot/blob/934981c85c779610e15ce549846b0bea212b283e/bot/build_error.go#L5-L9
+ */
+type ErrorType = 'none' | 'user' | 'system'
+
+/**
+ * The response we get back from the Buildbot deploy server, see:
+ * - https://github.com/netlify/buildbot/blob/934981c85c779610e15ce549846b0bea212b283e/deployserver/deploy_server.go#L20-L23
+ * - https://github.com/netlify/buildbot/blob/934981c85c779610e15ce549846b0bea212b283e/deployserver/deploy_server.go#L123-L129
+ */
+type BuildbotResponse = {
+  succeeded: boolean
+  values?: {
+    error?: string
+    code?: string
+    error_type?: ErrorType
+  }
+}
+
+/**
+ * Actions supported by the deploy server.
+ */
+enum Action {
+  /**
+   * Initiate the deploy.
+   */
+  DeploySite = 'deploySite',
+  /**
+   * Initiate the deploy, wait and hand back control to `@netlify/build` once the site is live.
+   */
+  DeploySiteAndWait = 'deploySiteAndAwaitLive',
+}
+
+/**
+ * Request format for the Buildbot deploy server, see:
+ * - https://github.com/netlify/buildbot/blob/934981c85c779610e15ce549846b0bea212b283e/deployserver/deploy_server.go#L15-L18
+ */
+type BuildRequest = {
+  action: Action
+  deployDir?: string
+}
+
+/**
+ * Creates the Buildbot IPC client we use to initiate the deploy
+ */
+export const createBuildbotClient = function (buildbotServerSocket: string) {
+  const connectionOpts = getConnectionOpts(buildbotServerSocket)
+  const client = net.createConnection(connectionOpts)
+  return client
+}
+
+/**
+ * Windows does not support Unix sockets well, so we also support `host:port`
+ */
+const getConnectionOpts = function (buildbotServerSocket: string): NetConnectOpts {
+  if (!buildbotServerSocket.includes(':')) {
+    return { path: buildbotServerSocket }
+  }
+
+  const [host, port] = buildbotServerSocket.split(':')
+  return { host, port: Number(port) }
+}
```

Comment:
This could probably be simplified since we don't need to support `host:port` syntax let alone Windows for this particular logic. Nevertheless left this untouched for now.